File: app/src/main/java/org/metabrainz/mobile/presentation/features/release_group/ReleaseGroupInfoFragment.java
package org.metabrainz.mobile.presentation.features.release_group;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import org.metabrainz.mobile.data.sources.api.entities.EntityUtils;
import org.metabrainz.mobile.data.sources.api.entities.WikiSummary;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.ReleaseGroup;
import org.metabrainz.mobile.databinding.CardReleaseGroupInfoBinding;

public class ReleaseGroupInfoFragment extends Fragment {

    private CardReleaseGroupInfoBinding binding;
    private ReleaseGroupViewModel releaseGroupViewModel;

    public static ReleaseGroupInfoFragment newInstance() {
        return new ReleaseGroupInfoFragment();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = CardReleaseGroupInfoBinding.inflate(inflater, container, false);

        releaseGroupViewModel = new ViewModelProvider(requireActivity()).get(ReleaseGroupViewModel.class);
        releaseGroupViewModel.getData().observe(getViewLifecycleOwner(), this::setReleaseGroupInfo);
        releaseGroupViewModel.getWikiData().observe(getViewLifecycleOwner(), this::setWiki);

        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    private void setWiki(WikiSummary wiki) {
        if (wiki != null) {
            String wikiText = wiki.getExtract();
            if (wikiText != null && !wikiText.isEmpty()) {
                showWikiCard();
                binding.wikiSummary.setText(wikiText);
            } else hideWikiCard();
        } else hideWikiCard();
    }

    private void showWikiCard() {
        binding.cardView.setVisibility(View.VISIBLE);
    }

    private void hideWikiCard() {
        binding.cardView.setVisibility(View.GONE);
    }

    private void setReleaseGroupInfo(ReleaseGroup releaseGroup) {
        if (releaseGroup != null) {
            String title, artist;
            title = releaseGroup.getTitle();
            artist = EntityUtils.getDisplayArtist(releaseGroup.getArtistCredits());
            binding.releaseGroupTitle.setText(title);
            if (artist != null && !artist.isEmpty())
                binding.releaseGroupArtist.setText("( ".concat(artist).concat(" )"));
            else
                binding.releaseGroupArtist.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/org/metabrainz/mobile/presentation/features/release_group/ReleaseGroupInfoFragment.kt
package org.metabrainz.mobile.presentation.features.release_group

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import org.metabrainz.mobile.data.sources.api.entities.EntityUtils
import org.metabrainz.mobile.data.sources.api.entities.WikiSummary
import org.metabrainz.mobile.data.sources.api.entities.mbentity.ReleaseGroup
import org.metabrainz.mobile.databinding.CardReleaseGroupInfoBinding
import org.metabrainz.mobile.presentation.features.base.MusicBrainzFragment
import org.metabrainz.mobile.util.Resource

class ReleaseGroupInfoFragment : Fragment() {
    private var binding: CardReleaseGroupInfoBinding? = null
    private val releaseGroupViewModel: ReleaseGroupViewModel by activityViewModels()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = CardReleaseGroupInfoBinding.inflate(inflater, container, false)
        releaseGroupViewModel.data.observe(viewLifecycleOwner) { setReleaseGroupInfo(it) }
        releaseGroupViewModel.wikiData.observe(viewLifecycleOwner) { setWiki(it) }
        return binding!!.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    private fun setWiki(wikiSummaryResource: Resource<WikiSummary>?) {
        if (wikiSummaryResource != null && wikiSummaryResource.status == Resource.Status.SUCCESS) {
            val wiki = wikiSummaryResource.data
            val wikiText = wiki.extract
            if (wikiText != null && !wikiText.isEmpty()) {
                showWikiCard()
                binding!!.wikiSummary.text = wikiText
            } else hideWikiCard()
        } else hideWikiCard()
    }

    private fun showWikiCard() {
        binding!!.cardView.visibility = View.VISIBLE
    }

    private fun hideWikiCard() {
        binding!!.cardView.visibility = View.GONE
    }

    private fun setReleaseGroupInfo(resource: Resource<ReleaseGroup>?) {
        if (resource != null && resource.status == Resource.Status.SUCCESS) {
            val releaseGroup = resource.data
            binding!!.releaseGroupTitle.text = releaseGroup.title
            if (releaseGroup.artistCredits != null)
                binding!!.releaseGroupArtist.text = EntityUtils.getDisplayArtist(releaseGroup.artistCredits)
            else binding!!.releaseGroupArtist.visibility = View.GONE
        }
    }

    companion object : MusicBrainzFragment {
        override fun newInstance(): Fragment {
            return ReleaseGroupInfoFragment()
        }
    }
}